src/languages/makefile.js

```diff
@@ -1,50 +1,89 @@
 /*
 Language: Makefile
 Author: Ivan Sagalaev <maniac@softwaremaniacs.org>
+Contributors: Joël Porquet <joel@porquet.org>
 Category: common
 */
 
 function(hljs) {
+  /* Variables: simple (eg $(var)) and special (eg $@) */
   var VARIABLE = {
     className: 'variable',
-    begin: /\$\(/, end: /\)/,
-    contains: [hljs.BACKSLASH_ESCAPE]
-  };
-  return {
-    aliases: ['mk', 'mak'],
-    contains: [
-      hljs.HASH_COMMENT_MODE,
-      {
-        begin: /^\w+\s*\W*=/, returnBegin: true,
-        relevance: 0,
-        starts: {
-          end: /\s*\W*=/, excludeEnd: true,
-          starts: {
-            end: /$/,
-            relevance: 0,
-            contains: [
-              VARIABLE
-            ]
-          }
-        }
-      },
+    variants: [
       {
-        className: 'section',
-        begin: /^[\w]+:\s*$/
+        begin: '\\$\\(' + hljs.IDENT_RE + '\\)',
+        contains: [hljs.BACKSLASH_ESCAPE],
       },
       {
-        className: 'meta',
-        begin: /^\.PHONY:/, end: /$/,
-        keywords: {'meta-keyword': '.PHONY'}, lexemes: /[\.\w]+/
+        begin: /\$[@%<?\^\+\*]/
       },
+    ]
+  };
+  /* Quoted string with variables inside */
+  var QUOTE_STRING = {
+    className: 'string',
+    begin: /"/, end: /"/,
+    contains: [
+      hljs.BACKSLASH_ESCAPE,
+      VARIABLE,
+    ]
+  };
+  /* Function: $(func arg,...) */
+  var FUNC = {
+    className: 'variable',
+    begin: /\$\([\w-]+\s/, end: /\)/,
+    keywords: {
+      built_in:
+        'subst patsubst strip findstring filter filter-out sort ' +
+        'word wordlist firstword lastword dir notdir suffix basename ' +
+        'addsuffix addprefix join wildcard realpath abspath error warning ' +
+        'shell origin flavor foreach if or and call eval file value',
+    },
+    contains: [
+      VARIABLE,
+    ]
+  };
+  /* Variable assignment */
+  var VAR_ASSIG = {
+    begin: '^' + hljs.IDENT_RE + '\\s*(:=|=|\\+=)',
+    illegal: '\\n',
+    returnBegin: true,
+    contains: [
       {
-        begin: /^\t+/, end: /$/,
-        relevance: 0,
-        contains: [
-          hljs.QUOTE_STRING_MODE,
-          VARIABLE
-        ]
+        className: 'variable',
```

Comment:
We usually highlight things as "variable" when they are a part of a string or, more generally, appear in a context where they should be distinguished from their surroundings. Here it feels to me like highlighting the left sign of the assignment might be unnecessary. I'll let you make the final call, but I'm -0 on having the class name. (The mode itself should stay even without a class name since to improve detection.)